File: sdk/lx-trading-py/tests/test_types.py
"""Tests for LX Trading SDK types module."""

import pytest
from decimal import Decimal

from lx_trading.types import (
    TradingPair,
    Balance,
    AggregatedBalance,
    OrderRequest,
    Order,
    Trade,
    Ticker,
    PriceLevel,
    Fee,
    Side,
    OrderType,
    OrderStatus,
    TimeInForce,
)


class TestTradingPair:
    def test_parse_dash_separator(self):
        pair = TradingPair.from_symbol("BTC-USDC")
        assert pair is not None
        assert pair.base == "BTC"
        assert pair.quote == "USDC"

    def test_parse_slash_separator(self):
        pair = TradingPair.from_symbol("ETH/USD")
        assert pair is not None
        assert pair.base == "ETH"
        assert pair.quote == "USD"

    def test_parse_underscore_separator(self):
        pair = TradingPair.from_symbol("SOL_USDT")
        assert pair is not None
        assert pair.base == "SOL"
        assert pair.quote == "USDT"

    def test_parse_invalid_symbol(self):
        pair = TradingPair.from_symbol("BTCUSDC")
        assert pair is None

    def test_to_hummingbot(self):
        pair = TradingPair(base="BTC", quote="USDC")
        assert pair.to_hummingbot() == "BTC-USDC"

    def test_to_ccxt(self):
        pair = TradingPair(base="ETH", quote="USD")
        assert pair.to_ccxt() == "ETH/USD"

    def test_str(self):
        pair = TradingPair(base="BTC", quote="USDC")
        assert str(pair) == "BTC-USDC"


class TestBalance:
    def test_total(self):
        balance = Balance(
            asset="BTC",
            venue="lx",
            free=Decimal("1.5"),
            locked=Decimal("0.5"),
        )
        assert balance.total == Decimal("2")


class TestAggregatedBalance:
    def test_total(self):
        balance = AggregatedBalance(
            asset="ETH",
            total_free=Decimal("10"),
            total_locked=Decimal("5"),
            by_venue=[],
        )
        assert balance.total == Decimal("15")


class TestOrderRequest:
    def test_market_order(self):
        order = OrderRequest.market("BTC-USDC", Side.BUY, Decimal("0.5"))
        assert order.symbol == "BTC-USDC"
        assert order.side == Side.BUY
        assert order.order_type == OrderType.MARKET
        assert order.quantity == Decimal("0.5")
        assert order.time_in_force == TimeInForce.IOC
        assert order.client_order_id  # Should have UUID

    def test_limit_order(self):
        order = OrderRequest.limit("ETH-USD", Side.SELL, Decimal("1"), Decimal("2000"))
        assert order.symbol == "ETH-USD"
        assert order.side == Side.SELL
        assert order.order_type == OrderType.LIMIT
        assert order.quantity == Decimal("1")
        assert order.price == Decimal("2000")
        assert order.time_in_force == TimeInForce.GTC

    def test_with_venue(self):
        order = OrderRequest.market("BTC-USDC", Side.BUY, Decimal("1"))
        order.with_venue("binance")
        assert order.venue == "binance"

    def test_with_post_only(self):
        order = OrderRequest.limit("BTC-USDC", Side.BUY, Decimal("1"), Decimal("50000"))
        order.with_post_only()
        assert order.post_only is True
        assert order.time_in_force == TimeInForce.POST_ONLY


class TestOrder:
    def test_is_open(self):
        order = Order(
            order_id="1",
            client_order_id="c1",
            symbol="BTC-USDC",
            venue="lx",
            side=Side.BUY,
            order_type=OrderType.LIMIT,
            status=OrderStatus.OPEN,
            quantity=Decimal("1"),
            filled_quantity=Decimal("0"),
            remaining_quantity=Decimal("1"),
            price=Decimal("50000"),
            average_price=None,
            created_at=1000,
            updated_at=1000,
        )
        assert order.is_open is True
        assert order.is_done is False

    def test_is_done(self):
        order = Order(
            order_id="2",
            client_order_id="c2",
            symbol="ETH-USD",
            venue="lx",
            side=Side.SELL,
            order_type=OrderType.MARKET,
            status=OrderStatus.FILLED,
            quantity=Decimal("1"),
            filled_quantity=Decimal("1"),
            remaining_quantity=Decimal("0"),
            price=None,
            average_price=Decimal("2000"),
            created_at=1000,
            updated_at=1000,
        )
        assert order.is_open is False
        assert order.is_done is True

    def test_fill_percent(self):
        order = Order(
            order_id="3",
            client_order_id="c3",
            symbol="BTC-USDC",
            venue="lx",
            side=Side.BUY,
            order_type=OrderType.LIMIT,
            status=OrderStatus.PARTIALLY_FILLED,
            quantity=Decimal("10"),
            filled_quantity=Decimal("3"),
            remaining_quantity=Decimal("7"),
            price=Decimal("50000"),
            average_price=Decimal("50000"),
            created_at=1000,
            updated_at=1000,
        )
        assert order.fill_percent == Decimal("30")


class TestTrade:
    def test_value(self):
        trade = Trade(
            trade_id="t1",
            order_id="o1",
            symbol="BTC-USDC",
            venue="lx",
            side=Side.BUY,
            price=Decimal("50000"),
            quantity=Decimal("0.1"),
            fee=Fee(asset="USDC", amount=Decimal("5")),
            timestamp=1000,
            is_maker=False,
        )
        assert trade.value == Decimal("5000")


class TestTicker:
    def test_mid_price_with_bid_ask(self):
        ticker = Ticker(
            symbol="BTC-USDC",
            venue="lx",
            bid=Decimal("49900"),
            ask=Decimal("50100"),
            last=Decimal("50000"),
            volume_24h=None,
            high_24h=None,
            low_24h=None,
            change_24h=None,
            timestamp=1000,
        )
        assert ticker.mid_price == Decimal("50000")

    def test_mid_price_without_bid_ask(self):
        ticker = Ticker(
            symbol="BTC-USDC",
            venue="lx",
            bid=None,
            ask=None,
            last=Decimal("50000"),
            volume_24h=None,
            high_24h=None,
            low_24h=None,
            change_24h=None,
            timestamp=1000,
        )
        assert ticker.mid_price == Decimal("50000")

    def test_spread(self):
        ticker = Ticker(
            symbol="BTC-USDC",
            venue="lx",
            bid=Decimal("49900"),
            ask=Decimal("50100"),
            last=Decimal("50000"),
            volume_24h=None,
            high_24h=None,
            low_24h=None,
            change_24h=None,
            timestamp=1000,
        )
        assert ticker.spread == Decimal("200")

    def test_spread_percent(self):
        ticker = Ticker(
            symbol="BTC-USDC",
            venue="lx",
            bid=Decimal("49900"),
            ask=Decimal("50100"),
            last=Decimal("50000"),
            volume_24h=None,
            high_24h=None,
            low_24h=None,
            change_24h=None,
            timestamp=1000,
        )
        # spread_percent = (200/49900) * 100 ≈ 0.401
        assert ticker.spread_percent is not None
        assert 0.4 < float(ticker.spread_percent) < 0.41


class TestPriceLevel:
    def test_value(self):
        level = PriceLevel(price=Decimal("100"), quantity=Decimal("5"))
        assert level.value == Decimal("500")

File: sdk/lx-trading-cpp/tests/test_types.cpp
// LX Trading SDK - Types Tests

#include <catch2/catch_test_macros.hpp>
#include <catch2/catch_approx.hpp>
#include <lx/trading/types.hpp>

using namespace lx::trading;
using Catch::Approx;

TEST_CASE("Decimal arithmetic", "[types]") {
    SECTION("Basic operations") {
        Decimal a = Decimal::from_double(100.5);
        Decimal b = Decimal::from_double(50.25);

        REQUIRE((a + b).to_double() == Approx(150.75));
        REQUIRE((a - b).to_double() == Approx(50.25));
        REQUIRE((a * Decimal::from_double(2.0)).to_double() == Approx(201.0));
        REQUIRE((a / Decimal::from_double(2.0)).to_double() == Approx(50.25));
    }

    SECTION("String conversion") {
        auto d = Decimal::from_string("123.456");
        REQUIRE(d.to_double() == Approx(123.456));

        auto d2 = Decimal::from_string("-99.99");
        REQUIRE(d2.to_double() == Approx(-99.99));
        REQUIRE(d2.is_negative());
    }

    SECTION("Comparison") {
        Decimal a = Decimal::from_double(10.0);
        Decimal b = Decimal::from_double(20.0);

        REQUIRE(a < b);
        REQUIRE(b > a);
        REQUIRE(a <= a);
        REQUIRE(a == a);
        REQUIRE(a != b);
    }

    SECTION("Zero and one") {
        REQUIRE(Decimal::zero().is_zero());
        REQUIRE(Decimal::one().to_double() == Approx(1.0));
    }
}

TEST_CASE("TradingPair parsing", "[types]") {
    SECTION("Hyphen separator") {
        auto pair = TradingPair::from_symbol("BTC-USDC");
        REQUIRE(pair.has_value());
        REQUIRE(std::string(pair->base.data()) == "BTC");
        REQUIRE(std::string(pair->quote.data()) == "USDC");
    }

    SECTION("Slash separator") {
        auto pair = TradingPair::from_symbol("ETH/USD");
        REQUIRE(pair.has_value());
        REQUIRE(std::string(pair->base.data()) == "ETH");
        REQUIRE(std::string(pair->quote.data()) == "USD");
    }

    SECTION("Underscore separator") {
        auto pair = TradingPair::from_symbol("LUX_USDT");
        REQUIRE(pair.has_value());
        REQUIRE(std::string(pair->base.data()) == "LUX");
        REQUIRE(std::string(pair->quote.data()) == "USDT");
    }

    SECTION("Invalid symbol") {
        auto pair = TradingPair::from_symbol("INVALID");
        REQUIRE_FALSE(pair.has_value());
    }

    SECTION("Format conversions") {
        auto pair = TradingPair::from_symbol("BTC-USDC");
        REQUIRE(pair->to_hummingbot() == "BTC-USDC");
        REQUIRE(pair->to_ccxt() == "BTC/USDC");
    }
}

TEST_CASE("OrderRequest factory methods", "[types]") {
    SECTION("Market order") {
        auto req = OrderRequest::market("BTC-USDC", Side::Buy, Decimal::from_double(1.5));
        REQUIRE(req.symbol == "BTC-USDC");
        REQUIRE(req.side == Side::Buy);
        REQUIRE(req.order_type == OrderType::Market);
        REQUIRE(req.quantity.to_double() == Approx(1.5));
        REQUIRE(req.time_in_force == TimeInForce::IOC);
    }

    SECTION("Limit order") {
        auto req = OrderRequest::limit("ETH-USDC", Side::Sell,
            Decimal::from_double(10.0), Decimal::from_double(2000.0));
        REQUIRE(req.symbol == "ETH-USDC");
        REQUIRE(req.side == Side::Sell);
        REQUIRE(req.order_type == OrderType::Limit);
        REQUIRE(req.price.value().to_double() == Approx(2000.0));
        REQUIRE(req.time_in_force == TimeInForce::GTC);
    }

    SECTION("Builder methods") {
        auto req = OrderRequest::market("BTC-USDC", Side::Buy, Decimal::from_double(1.0))
            .with_venue("lx_dex")
            .with_post_only()
            .with_client_id("my-order-123");

        REQUIRE(req.venue.value() == "lx_dex");
        REQUIRE(req.post_only == true);
        REQUIRE(req.time_in_force == TimeInForce::PostOnly);
        REQUIRE(req.client_order_id == "my-order-123");
    }
}

TEST_CASE("Order status checks", "[types]") {
    Order order;
    order.quantity = Decimal::from_double(100.0);
    order.filled_quantity = Decimal::from_double(50.0);
    order.remaining_quantity = Decimal::from_double(50.0);

    SECTION("Partially filled") {
        order.status = OrderStatus::PartiallyFilled;
        REQUIRE(order.is_open());
        REQUIRE_FALSE(order.is_done());
        REQUIRE(order.fill_percent().to_double() == Approx(50.0));
    }

    SECTION("Filled") {
        order.status = OrderStatus::Filled;
        order.filled_quantity = order.quantity;
        REQUIRE_FALSE(order.is_open());
        REQUIRE(order.is_done());
    }

    SECTION("Cancelled") {
        order.status = OrderStatus::Cancelled;
        REQUIRE_FALSE(order.is_open());
        REQUIRE(order.is_done());
    }
}

TEST_CASE("Ticker calculations", "[types]") {
    Ticker ticker;
    ticker.bid = Decimal::from_double(100.0);
    ticker.ask = Decimal::from_double(101.0);

    SECTION("Mid price") {
        auto mid = ticker.mid_price();
        REQUIRE(mid.has_value());
        REQUIRE(mid->to_double() == Approx(100.5));
    }

    SECTION("Spread") {
        auto spread = ticker.spread();
        REQUIRE(spread.has_value());
        REQUIRE(spread->to_double() == Approx(1.0));
    }

    SECTION("Spread percent") {
        auto spread_pct = ticker.spread_percent();
        REQUIRE(spread_pct.has_value());
        REQUIRE(spread_pct->to_double() == Approx(1.0));
    }
}

TEST_CASE("Enum to_string", "[types]") {
    REQUIRE(std::string(to_string(Side::Buy)) == "buy");
    REQUIRE(std::string(to_string(Side::Sell)) == "sell");
    REQUIRE(std::string(to_string(OrderType::Market)) == "market");
    REQUIRE(std::string(to_string(OrderType::Limit)) == "limit");
    REQUIRE(std::string(to_string(TimeInForce::GTC)) == "GTC");
    REQUIRE(std::string(to_string(OrderStatus::Filled)) == "filled");
}
